README.rst
Pantry: LXC-Web-Panel reloaded
==========================

.. image:: https://travis-ci.org/EstudioNexos/LXC-Web-Panel.svg?branch=master
    :target: https://travis-ci.org/EstudioNexos/LXC-Web-Panel

Pantry is an easy to use Linux Containers manager.

- NEW Installer to make it even easier to start running.
- Display some host and container information: IP, CPU, Mem and Disk usage.
- List, create, delete containers and change container settings.
- Start, stop and freeze containers
- Create container backup to a tar.gz
- List, create and delete snapshots
- Show system compatibility
- Manage panel users
- Expose container operations through a REST API and Manage API authoritation tokens

.. image:: https://github.com/EstudioNexos/LXC-Web-Panel/raw/master/screenshots/dashboard.png
  :width: 400
  :alt: Revamped dashboard


This is a fork of the original LXC-Web-Panel from https://github.com/lxc-webpanel/LXC-Web-Panel and https://github.com/claudyus/LXC-Web-Panel/ looking for LXC 2.x compatibility, Python 3, UI updating, New features like snapshots and backup restore management, and Fabric (fabfile) integration to be able to manage remote hosts.

The code was tested on Debian 10 Buster and Python 3.8.

We are working on this heavily so expect frequent changes in code, installation docs outdated, ...

All contributions are welcomed.

.. image:: screenshots/container_details.png
  :width: 300
  :alt: Container snapshots

.. image:: screenshots/create_user.png
  :width: 300
  :alt: Container snapshots

.. image:: screenshots/container_snapshots.png
  :width: 300
  :alt: Container snapshots

Installation from source code
----------------------------------------------

Easiest way to run it is using Pyenv:

::
  curl https://pyenv.run | bash

Add following to your .bashrc or .zshrc

::
  export PATH="$HOME/.pyenv/bin:$PATH"
  eval "$(pyenv init -)"
  eval "$(pyenv virtualenv-init -)"

After that continue Pyenv setup:

::
  exec $SHELL
  pyenv update
  pyenv install 3.8.1
  pyenv rehash
  pyenv global 3.8.1 OR pyenv local 3.8.1



::
  git clone https://github.com/EstudioNexos/LXC-Web-Panel.git pantry
  cd pantry
  pip install -r requirements.txt
  python setup.py install
  ./bin/gtr        # run lwp wth debug support

We recomend using /var/pantry/backups or /var/backups/pantry path.

First run we will get an installer page, most of defaults are OK and just click INSTALL.

Then stop gtr with CTRL+C and start it again.

Default login is admin/admin but soon it will be configurable through the installer.

Your lwp panel is now at http://localhost:5000/.

htpasswd
++++++++

To enable authentication against htpasswd file you should set ``auth`` type to ``htpasswd`` and ``file`` variable in ``htpasswd`` section to point to the htpasswd file.

This backend use the crypt function, here an example where ``-d`` force the use of crypt encryption when generating the htpasswd file::

  htpasswd -d -b -c /etc/lwp/httpasswd admin admin

PAM
+++

To enable authentication against PAM you should set ``auth`` type to ``pam`` and ``service`` variable in ``pam`` section.
Python PAM module needs to be installed::

  apt-get install python-pam

or

::

  pip install pam

or

::

  yum install python-pam

With default ``login`` service all valid linux users can login to pantry.
Many more options are available via PAM Configuration, see PAM docs.

HTTP
+++++

This auth method is used to authenticate the users using an external http server through a POST request. To enable this method  ``auth`` type to ``http`` and configure the option under ``http`` section.


LICENSE
-------
This work is released under MIT License, see LICENSE file.


CHANGELOG.rst
Version 0.11-beta2 (07-02-2020)
===========================================================

More cleanup (should have started from scratch?)
Rename to pantry

Version 0.10-beta1 (24-09-2018)
===========================================================

Removed Features
-----------------------

* Python-ldap is not a requierment for setup.py install

Bug Fixes
-----------------------
* lxc-clone replaced by lxc-copy

Added Features
-----------------------

* Updated UI libraries (Bootstrap 4, FontAwesome 4)
* New theme Bootswatch
* List snapshots (in progress)
* Reorganized dashboard and container views

Version 0.9 (31-01-2016)
===========================================================

Added Features
-----------------------

* Display boot flag in main view
* Add auth method
* Improving test
* Added systemd support
* Add loop backing storage option

Bug Fixes
-----------------------

* Rename generate session method
* Fix regexs used in edit form
* Fix regex to allow creation of container of appropriate size over LVM backend

Thanks to Lajos Santa, Lukáš Raška, Marcos Silva Cunha, Marton Suranyi, Sebastian Reimers and Yoav Steinberg for their contributions to this release.


Other
------------------------

Since this release you are encuraged to use the packagecloud.io repo instead of github repo.


Version 0.8 (12-02-2015)
===========================================================

TODO

Thanks to Alexander Knöbel, Boscorillium, Claudio Cesar Sanchez Tejeda,
Claudio Valenti and Lukáš Raška for their contributions to this release.


lwp/installer.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
from __future__ import absolute_import, print_function

import os, sys, socket
import subprocess
import string, random
def id_generator(size=24, chars=string.ascii_uppercase + string.digits):
    return ''.join(random.choice(chars) for _ in range(size))

from flask import Flask, request, session, g, redirect, url_for, abort, render_template
from lwp.utils import ConfigParser
#~ from lwp.utils import connect_db, check_session_limit, config
#~ import lwp.lxclite as lxc

DEBUG = False
ADDRESS = 'localhost'
PORT = 5000
PREFIX = '/install'

# Flask app
app = Flask('lwp', static_url_path="{0}/static".format(PREFIX))
app.config.from_object(__name__)

def create(path):
    if os.path.exists(path) == False:
        os.makedirs(path)
    return path

def get_parent_path(path):
    sp = path.split('/')
    parent = os.path.join('/'.join(sp[0:-1]))
    return parent

def is_already_installed():
    try:
        from lwp.config import read_config_file
        config = read_config_file()
        config.get('database', 'file')
        return True
    except: pass
    return False

@app.route('/', methods=['POST', 'GET'], defaults={'path': ''})
@app.route('/<path:path>', methods=['POST', 'GET'])
def install(path):
    """
    Installer
    """ 
    exec_path = os.path.abspath(os.path.dirname(sys.argv[0]))
    if len(path) > 0:
        return redirect('/')
    test_file = os.path.join('/etc',id_generator())
    can_we_install = False
    already_installed = is_already_installed()
    #~ print(already_installed)
    context = {
        'can_we_install': can_we_install,
        'already_installed': already_installed,
    }
    

    try:
        open(test_file, 'w').close()
    except:
        pass
    if os.path.exists(test_file):
        context['can_we_install'] = True
        os.remove(test_file)
    else:
        return render_template('installer.html', **context)
    context['checks'] = {}#lxc.checkconfig()
    if request.method == 'POST':
        from lwp.utils import hash_passwd
        f = request.form
        datadir = f.get('datadir','/var/lwp')
        create(datadir)
        conffile = f.get('conffile','/etc/lwp/lwp.conf')
        create(get_parent_path(conffile))
        config = ConfigParser()
        config['global'] = {}
        config['global']['address'] = f.get('address','127.0.0.1')
        config['global']['debug'] = f.get('debug','False')
        config['global']['port'] = f.get('port','5000')
        config['global']['auth'] = f.get('auth','database')
        config['global']['prefix'] = f.get('prefix','')
        config['storage_repository'] = {}
        config['storage_repository']['local'] = f.get('local_storage_repository','backups')
        create(os.path.join(datadir,config['storage_repository']['local']))
        config['database'] = {}
        config['database']['file'] = f.get('database_uri','sqlite:////var/lwp/lwp.db')
        config['session'] = {}
        config['session']['time'] = f.get('time','10')
        config['api'] = {}
        internal_token = hash_passwd('lwp')
        config['api']['username'] = f.get('api_username','admin')
        config['api']['token'] = f.get('api_token',internal_token)
        with open(conffile, 'w') as configfile:
            config.write(configfile)
        
        from lwp.database.models import get_database,Users,ApiTokens,Projects,Hosts,Containers,ContainerTag,Tags

        database = get_database()
        database.create_tables([Users,ApiTokens,Projects,Hosts,Containers,ContainerTag,Tags])
        admin = Users.create(name='Admin',username='admin',su='Yes',password=hash_passwd('admin'))
        ApiTokens.create(username='admin',description='internal',token=internal_token)
        host = Hosts.create(hostname=socket.gethostname(), admin=admin, api_token=config['api']['token'], api_user=config['api']['username'])
        default_project = Projects.create(title='Default',description='Default project to start with',admin=admin)
        subprocess.check_call('touch {}'.format(exec_path), shell=True)
        context['already_installed'] = is_already_installed()
    return render_template('installer.html', **context)
